generator/autorest.ts

```diff
@@ -2,15 +2,97 @@
+import { findRecursive, executeCmd, fileExists } from './utils';
 import * as constants from './constants';
-import { ReadmeTag, AutoGenConfig, CodeBlock } from './models';
-import * as cm from '@ts-common/commonmark-to-markdown'
-import * as yaml from 'js-yaml'
 import { readFile, writeFile } from 'fs/promises';
+import * as markdown from '@ts-common/commonmark-to-markdown'
+import * as yaml from 'js-yaml'
 
 const autorestBinary = os.platform() === 'win32' ? 'autorest.cmd' : 'autorest';
-export const apiVersionRegex = /^\d{4}-\d{2}-\d{2}(|-preview)$/;
+
+const rootDir = `${__dirname}/../`;
+const extensionDir = path.resolve(`${rootDir}/bicep-types-az/src/autorest.bicep/`);
+
+export async function generateAutorestConfig(readmePath: string, bicepReadmePath: string) {
+  // We expect a path format convention of <provider>/(any/number/of/intervening/folders)/<yyyy>-<mm>-<dd>(|-preview)/<filename>.json
+  // This information is used to generate individual tags in the generated autorest configuration
+  // eslint-disable-next-line no-useless-escape
+  const pathRegex = /^(\$\(this-folder\)\/|)([^\/]+)(?:\/[^\/]+)*\/(\d{4}-\d{2}-\d{2}(|-preview))\/.*\.json$/i;
+
+  const readmeContents = await readFile(readmePath, { encoding: 'utf8' });
+  const readmeMarkdown = markdown.parse(readmeContents);
+
+  const inputFiles = new Set<string>();
+  // we need to look for all autorest configuration elements containing input files, and collect that list of files. These will look like (e.g.):
+  // ```yaml $(tag) == 'someTag'
+  // input-file:
+  // - path/to/file.json
+  // - path/to/other_file.json
+  // ```
+  for (const node of markdown.iterate(readmeMarkdown.markDown)) {
+    // We're only interested in yaml code blocks
+    if (node.type !== 'code_block' || !node.info || !node.literal ||
+      !node.info.trim().startsWith('yaml')) {
+      continue;
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const yamlData = yaml.load(node.literal) as any;
+    if (yamlData) {
+      // input-file may be a single string or an array of strings
+      const inputFile = yamlData['input-file'];
+      if (typeof inputFile === 'string') {
+        inputFiles.add(inputFile);
+      } else if (inputFile instanceof Array) {
+        for (const i of inputFile) {
+          inputFiles.add(i);
+        }
+      }
+    }
+  }
+
+  const filesByTag: Record<string, string[]> = {};
+  for (const file of inputFiles) {
+    const normalizedFile = normalizeJsonPath(file);
+    const match = pathRegex.exec(normalizedFile);
+    if (match) {
+      // Generate a unique tag. We can't process all of the different API versions in one autorest pass
+      // because there are constraints on naming uniqueness (e.g. naming of definitions), so we want to pass over
+      // each API version separately.
+      const tagName = `${match[2].toLowerCase()}-${match[3].toLowerCase()}`;
```

Comment:
Not fully following this, I understand that we want to generate something unique but the why is not clear, is it because autorest chokes? because our code chokes? because the generated schema will have duplicates?

Also, what is the value of `match[2]` and `match[3]` how do you ensure these are unique?

Reply:
Let me add a comment to explain better. Autorest groups files by "tag", and we're generating the config for autorest here. So the tags just need to be unique for grouping.

Here's an example of what the output format looks like:

    ##Bicep

    ### Bicep multi-api
    ```yaml $(bicep) && $(multiapi)
    batch:
      - tag: microsoft.securityinsights-2024-03-01
      - tag: microsoft.securityinsights-2024-01-01-preview
      ...

    ```
    ### Tag: microsoft.securityinsights-2024-03-01 and bicep
    ```yaml $(tag) == 'microsoft.securityinsights-2024-03-01' && $(bicep)
    input-file:
      - Microsoft.SecurityInsights/stable/2024-03-01/AlertRules.json
      - Microsoft.SecurityInsights/stable/2024-03-01/AutomationRules.json
      ...

    ```
    ### Tag: microsoft.securityinsights-2024-01-01-preview and bicep
    ```yaml $(tag) == 'microsoft.securityinsights-2024-01-01-preview' && $(bicep)
    input-file:
      - Microsoft.SecurityInsights/preview/2024-01-01-preview/AlertRules.json
      - Microsoft.SecurityInsights/preview/2024-01-01-preview/AutomationRules.json
      - Microsoft.SecurityInsights/preview/2024-01-01-preview/BillingStatistics.json
      ...